File: spike/bench/bench.py
#!/usr/bin/env python3
"""Latency harness (Python side): FFI vs WASM vs pure-Python vs subprocess.

Warm: median of N calls in one process over bench/tree.
Cold: median of fresh-process runs (includes interpreter/runtime startup).
Run inside spike/.venv after `maturin develop --release` and the wasm build.
"""

import hashlib
import json
import statistics
import subprocess
import sys
import time
from pathlib import Path

SPIKE = Path(__file__).resolve().parents[1]
TREE = str(SPIKE / "bench" / "tree")
NATIVE = str(SPIKE / "target" / "release" / "agent-context")
GLOBS = ["**/*.md"]
WARM_N = 1000
COLD_N = 20

sys.path.insert(0, str(SPIKE / "bench"))
sys.path.insert(0, str(SPIKE / "wasm" / "py"))


def median_ms(fn, n):
    times = []
    for _ in range(n):
        t0 = time.perf_counter()
        fn()
        times.append((time.perf_counter() - t0) * 1000)
    return statistics.median(times)


def sha(text):
    return hashlib.sha256(text.encode()).hexdigest()[:12]


def main():
    results = {}

    # --- pure-Python baseline ---
    from baseline_pure import print_tree
    results["pure-py warm"] = median_ms(lambda: print_tree(GLOBS, TREE), WARM_N)
    pure = print_tree(GLOBS, TREE)

    # --- subprocess-over-native-binary baseline ---
    def sub():
        return subprocess.run(
            [NATIVE, "print", *GLOBS, "--cwd", TREE],
            capture_output=True, check=True, text=True,
        ).stdout
    results["subprocess warm=cold"] = median_ms(sub, COLD_N)
    sub_text = sub()

    # --- FFI (pyo3) ---
    import agent_context as ffi
    results["ffi warm"] = median_ms(lambda: ffi.print(GLOBS, TREE), WARM_N)
    ffi_r = ffi.print(GLOBS, TREE)
    cold_cmd = (
        f"import agent_context; agent_context.print({GLOBS!r}, {TREE!r})"
    )
    results["ffi cold (fresh python)"] = median_ms(
        lambda: subprocess.run([sys.executable, "-c", cold_cmd], check=True),
        COLD_N,
    )

    # --- WASM (wasmtime-py) ---
    from sdk import AgentContext
    ac = AgentContext(TREE)
    results["wasm warm"] = median_ms(lambda: ac.print(GLOBS), WARM_N)
    wasm_r = ac.print(GLOBS)
    results["wasm re-instantiate+call (in-process)"] = median_ms(
        lambda: AgentContext(TREE).print(GLOBS), COLD_N
    )
    wasm_cold = (
        f"import sys; sys.path.insert(0, {str(SPIKE / 'wasm' / 'py')!r}); "
        f"from sdk import AgentContext; AgentContext({TREE!r}).print({GLOBS!r})"
    )
    results["wasm cold (fresh python)"] = median_ms(
        lambda: subprocess.run([sys.executable, "-c", wasm_cold], check=True),
        COLD_N,
    )

    # --- interpreter-startup floor for cold context ---
    results["python noop (fresh python)"] = median_ms(
        lambda: subprocess.run([sys.executable, "-c", "pass"], check=True),
        COLD_N,
    )

    print(json.dumps({k: round(v, 3) for k, v in results.items()}, indent=2))
    print("parity sha256/12:", {
        "pure-py": sha(pure["text"]),
        "subprocess": sha(sub_text),
        "ffi": sha(ffi_r.text),
        "wasm": sha(wasm_r.text),
    })


if __name__ == "__main__":
    main()

File: spike/bench/bench.mjs
#!/usr/bin/env node
// Latency harness (Node side): FFI vs WASM vs subprocess.
// Run from spike/ after `napi build --platform --release` and the wasm build.
import { execFileSync } from 'node:child_process';
import { createHash } from 'node:crypto';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const SPIKE = dirname(dirname(fileURLToPath(import.meta.url)));
const TREE = join(SPIKE, 'bench', 'tree');
const NATIVE = join(SPIKE, 'target', 'release', 'agent-context');
const GLOBS = ['**/*.md'];
const WARM_N = 1000;
const COLD_N = 20;

const medianMs = (fn, n) => {
  const times = [];
  for (let i = 0; i < n; i++) {
    const t0 = process.hrtime.bigint();
    fn();
    times.push(Number(process.hrtime.bigint() - t0) / 1e6);
  }
  return times.sort((a, b) => a - b)[Math.floor(n / 2)];
};

const sha = (text) => createHash('sha256').update(text).digest('hex').slice(0, 12);

const results = {};

// --- subprocess-over-native-binary baseline ---
const sub = () => execFileSync(NATIVE, ['print', ...GLOBS, '--cwd', TREE], { encoding: 'utf8' });
results['subprocess warm=cold'] = medianMs(sub, COLD_N);
const subText = sub();

// --- FFI (napi-rs) ---
const ffi = await import(join(SPIKE, 'ffi', 'node', 'index.js')).then(m => m.default ?? m);
results['ffi warm'] = medianMs(() => ffi.print(GLOBS, TREE), WARM_N);
const ffiR = ffi.print(GLOBS, TREE);
const ffiCold = `const ffi = require(${JSON.stringify(join(SPIKE, 'ffi', 'node', 'index.js'))}); ffi.print(${JSON.stringify(GLOBS)}, ${JSON.stringify(TREE)})`;
results['ffi cold (fresh node)'] = medianMs(
  () => execFileSync(process.execPath, ['-e', ffiCold]), COLD_N);

// --- WASM (jco) ---
// NOTE: the jco-transpiled module instantiates once at import time (module
// singleton) — there is no supported in-process re-instantiation, so cold
// for jco is a fresh node process.
const { sandbox, print } = await import(join(SPIKE, 'wasm', 'node', 'sdk.mjs'));
const guest = sandbox(TREE);
results['wasm warm'] = medianMs(() => print(GLOBS, guest), WARM_N);
const wasmR = print(GLOBS, guest);
const wasmCold = `import(${JSON.stringify(join(SPIKE, 'wasm', 'node', 'sdk.mjs'))}).then(s => s.print(${JSON.stringify(GLOBS)}, s.sandbox(${JSON.stringify(TREE)})))`;
results['wasm cold (fresh node)'] = medianMs(
  () => execFileSync(process.execPath, ['-e', wasmCold]), COLD_N);

// --- runtime-startup floor for cold context ---
results['node noop (fresh node)'] = medianMs(
  () => execFileSync(process.execPath, ['-e', '0']), COLD_N);

console.log(JSON.stringify(Object.fromEntries(
  Object.entries(results).map(([k, v]) => [k, Math.round(v * 1000) / 1000])), null, 2));
console.log('parity sha256/12:', {
  subprocess: sha(subText),
  ffi: sha(ffiR.text),
  wasm: sha(wasmR.text),
});
